### ISSUE TITLE
Halt Condition and Vocabulary Size Discrepancy

### ISSUE THREAD

**Issue by sup3rgiu** [gerstung-lab/Delphi#2]
Hi,  

I'm trying to understand the halt condition in the `generate` function. Based on the following lines of code, my interpretation is that the loop could terminate at each iteration if either:  

1. The "next token" for all the samples is Death, **or**  
2. The "next age" exceeds the maximum age for all samples.  

After termination, all tokens (and corresponding ages) following the first death token are padded with the padding token (`0`).  

https://github.com/gerstung-lab/Delphi/blob/d4d0839cbffbb4b1e3689172b681c0ccc73c22bd/model.py#L415-L418 

Now, considering the simulated data in the `ukb_simulated_data` folder, it appears that the death token has a value of **1268** (the maximum value found in the `.bin` files). This value is then incremented by **1** (to accommodate the padding token), making it **1269**. This is consistent with the [labels.csv](https://github.com/gerstung-lab/Delphi/blob/main/data/ukb_simulated_data/labels.csv) file, which already accounts for padding and "no event" tokens — placing the death token at row **1270** (i.e., **1269** if zero-indexed).  

Now, my concern is the following:  

In [train_delphi_demo.py](https://github.com/gerstung-lab/Delphi/blob/main/config/train_delphi_demo.py), the vocabulary size is set to **1271**, not **1270** (which would be `max_token_id + 1`). As a result, in the previously mentioned code snippet, the death token is **never** matched with `vocab_size - 1`.  

### Minimal Reproducible Code  

```python
import torch

bs = 5
seq_len = 10
death_token_id = 1269

def generate_samples(vocab_size):

    torch.manual_seed(42)
    
    print(f"*** Vocab size: {vocab_size} ***")

    idx = torch.randint(0, vocab_size-1, (bs, seq_len)) # generate some random data
    idx[idx == death_token_id] = 20 # remove any death token that might have been generated

    dead_samples_idx = [0,1,2,3,4] # idx of samples that died

    idx[dead_samples_idx, 7] = death_token_id # insert a death token in the 8th position
    idx[dead_samples_idx[:2], 4] = death_token_id # in 2 samples, insert a death token in the 5th position

    pad = (torch.cumsum(torch.cumsum(idx == vocab_size -1,1).int(),1) > 1) #compute padding mask

    print(f"Non-padded sequences:\n{idx}")

    idx[pad] = 0

    print(f"\nPadded sequences:\n{idx}\n\n")
    
    return idx

_ = generate_samples(1271)
_ = generate_samples(1270)
```

Which for instance will print:

```
*** Vocab size: 1271 ***
Non-padded sequences:
tensor([[ 932,   67,  656,  644, 1269, 1175,  750, 1269, 1040,  963],
        [ 338,  744, 1110,  364, 1269,  182,  285, 1269,  217,  386],
        [1239,  356,  263,  841, 1009,  823,    1, 1269,  397,  739],
        [ 142,  620, 1125, 1039, 1153,  374,  409, 1269,  442,  690],
        [ 376,  572,  347, 1199,  597,  963, 1193, 1269,  873,  817]])

Padded sequences:
tensor([[ 932,   67,  656,  644, 1269, 1175,  750, 1269, 1040,  963],
        [ 338,  744, 1110,  364, 1269,  182,  285, 1269,  217,  386],
        [1239,  356,  263,  841, 1009,  823,    1, 1269,  397,  739],
        [ 142,  620, 1125, 1039, 1153,  374,  409, 1269,  442,  690],
        [ 376,  572,  347, 1199,  597,  963, 1193, 1269,  873,  817]])


*** Vocab size: 1270 ***
Non-padded sequences:
tensor([[1113, 1049,  196,  454, 1269, 1001,    7, 1269,  769, 1136],
        [1255,   22,  845,  225, 1269,  598,  236, 1269,  219, 1156],
        [1057,  782,    1,  986,  305, 1111, 1118, 1269, 1108, 1027],
        [1119,   84,  871, 1096,  476,  160,  569, 1269,  595,   75],
        [1109,   93,   28,  390,   39, 1016,  365, 1269,  135,  137]])

Padded sequences:
tensor([[1113, 1049,  196,  454, 1269,    0,    0,    0,    0,    0],
        [1255,   22,  845,  225, 1269,    0,    0,    0,    0,    0],
        [1057,  782,    1,  986,  305, 1111, 1118, 1269,    0,    0],
        [1119,   84,  871, 1096,  476,  160,  569, 1269,    0,    0],
        [1109,   93,   28,  390,   39, 1016,  365, 1269,    0,    0]])
```

As you can see, in the first case the tokens after the first Death token (1269) are not padded, while in the second case they are.

Would appreciate any clarification on whether this is a potential issue. Thanks!

**Comment by yozhikoff** [comment]
Hi @sup3rgiu, 

You are absolutely correct!
The vocab size should have been 1270 in this case. I will fix it later today.

That being said, perhaps the current way the Death token is identified by the `generate` function (aka the `vocab_size - 1`) is suboptimal anyway. It's quite handy to introduce new tokens by appending them to the end, which would, of course, break the current implementation. I should add `dealth_token_id` to the model config and use it there. I will do this soon as well.

Thanks for spotting and documenting this issue! Please feel free to get in touch if you have any other questions about Delphi.

**Comment by sup3rgiu** [comment]
I'm currently handling it in this way:

```python
    @torch.no_grad()
    def generate(self, idx, age, max_new_tokens=100, max_age = 85*365.25, temperature=1.0, top_k=None, no_repeat=True, death_tokens=None):
        """
        Take a conditioning sequence of indices idx (LongTensor of shape (b,t)) and complete
        the sequence max_new_tokens times, feeding the predictions back into the model each time.
        Most likely you'll want to make sure to be in model.eval() mode of operation for this.
        """
        if death_tokens is not None:
            if isinstance(death_tokens, int):
                death_tokens = [death_tokens]
            
            if isinstance(death_tokens, list):
                death_tokens = torch.tensor(death_tokens, device=idx.device)
            elif torch.is_tensor(death_tokens):
                death_tokens = death_tokens.to(idx.device)
            else:
                raise ValueError("death_tokens must be an int, list or tensor")
        else:
            death_tokens = [self.config.vocab_size -1]
        
        if max_new_tokens == -1:
            max_new_tokens = 10000
        for _ in range(max_new_tokens):
            # if the sequence context is growing too long we must crop it at block_size
            idx_cond = idx #if idx.size(1) <= self.config.block_size else idx[:, -self.config.block_size:]
            age_cond = age #if age.size(1) <= self.config.block_size else age[:, -self.config.block_size:]

            # forward the model to get the logits for the index in the sequence
            logits, _, _ = self(idx_cond, age_cond)
            # pluck the logits at the final step and scale by desired temperature
            logits = logits[:, -1, :] / temperature
            logits[:,self.config.ignore_tokens] = -float('Inf')
            # optionally crop the logits to only the top k options
            if top_k is not None:
                v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
                logits[logits < v[:, [-1]]] = -float('Inf')
                
            if no_repeat:
                fill = idx + 0
                fill[fill == 1] = 0
                logits = logits.scatter_(1, fill, -float("Inf"))
            
            # apply softmax to convert logits to (normalized) probabilities
            # probs = F.softmax(logits, dim=-1)
            # sample from the distribution
            # idx_next = torch.multinomial(probs, num_samples=1)
            # lse = torch.logsumexp(logits, -1)[:,None]
            t_next = torch.clamp( -torch.exp(-logits) * torch.rand(logits.shape, device=idx.device).log(),
                min=0, max=365*80.
            ).min(1)
            #age_next = age[...,[-1]] + torch.clamp(-torch.exp(-lse) * torch.rand(lse.shape, device=idx.device).log(), min=self.config.t_min, max=365*80.) #torch.normal(torch.zeros((1,1), device=idx.device),1.)
            idx_next = t_next[1][:,None]
            age_next = age[...,[-1]] + t_next[0][:,None]
            
            # append sampled index to the running sequence and continue
            idx = torch.cat((idx, idx_next), dim=1)
            age = torch.cat((age, age_next), dim=1)
            
            if torch.all(torch.isin(idx_next, death_tokens)) or torch.all(age_next > max_age):
                break
        
        pad = (torch.cumsum(torch.cumsum(torch.isin(idx, death_tokens), 1).int(), 1) > 1) + (age > max_age)
        logits, _, _ = self(idx, age)
        idx[pad] = 0
        age[pad] = float('NaN')
        if no_repeat:
            fill = idx + 0
            fill[fill == 1] = 0
            logits = torch.stack([logits[:,j].scatter_(1, fill[:,:j+1], float("NaN")) for j in range(fill.shape[1])]).transpose(0,1)

        return idx, age, logits
```

You can even define multiple death tokens this way (for example, if you want to define two causes of death).

**Comment by yozhikoff** [comment]
Good idea; I implemented it similarly. 

Thanks!

**Comment by HelloWorldLTY** [comment]
I think this question is related to the issue I proposed. It seems that the death label is still 1268 in the simulation data, but all the phenotypes have been shifted with -2 (for example, somking 10->8). https://github.com/gerstung-lab/Delphi/issues/27
